zboss/CMakeLists.txt

```diff
@@ -0,0 +1,70 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+include(${NRFXLIB_DIR}/common.cmake)
+
+#old_sdk: ToDo - Add include path for radio using Kconfig. Now isn't working on NCS master
+zephyr_include_directories(${NRFXLIB_DIR}/../modules/hal/nordic/drivers/nrf_radio_802154)
+
+# Common includes, not related to ZBOSS source code.
+zephyr_include_directories(src/zb_error)
+
+# Source files
+zephyr_sources(src/zb_error/zb_error_to_string.c)
+
+# Suppress bitfield compatibility warning
+zephyr_compile_options(-Wno-packed-bitfield-compat)
+
+# Add compile-time definition, indicating which ZBOSS library will be linked
+if (CONFIG_ZIGBEE_ROLE_END_DEVICE)
+  zephyr_compile_definitions(ZB_ED_ROLE)
+endif()
+
+
+if (CONFIG_ZBOSS_LIBRARY_PRECOMPILED)
+
+zephyr_include_directories(include)
+zephyr_include_directories(include/zcl)
+zephyr_include_directories(include/ha)
+zephyr_include_directories(include/se)
+zephyr_include_directories(include/osif)
+zephyr_include_directories(include/addons)
+
+nrfxlib_calculate_lib_path(lib_path)
+set(ZBOSS_LIB_PATH ${CMAKE_CURRENT_SOURCE_DIR}/${lib_path})
+if(NOT EXISTS ${ZBOSS_LIB_PATH})
+  message(WARNING "This combination of SoC and floating point ABI is not supported by the ZBOSS lib."
+                  "(${ZBOSS_LIB_PATH} doesn't exist.)")
+endif()
+
+if (CONFIG_ZIGBEE_ROLE_COORDINATOR OR CONFIG_ZIGBEE_ROLE_ROUTER)
+  zephyr_link_libraries(${ZBOSS_LIB_PATH}/libzboss.a)
+elseif (CONFIG_ZIGBEE_ROLE_END_DEVICE)
+  zephyr_link_libraries(${ZBOSS_LIB_PATH}/libzboss.ed.a)
+else()
+  message( FATAL_ERROR "Unsupported Zigbee role. Exiting." )
+endif()
+
+elseif(CONFIG_ZBOSS_LIBRARY_SOURCES)
+  # Read ZOI platform repository path from west manifest file.
+  execute_process(
+    COMMAND
+    ${WEST} list -f {posixpath} manifest
+    OUTPUT_VARIABLE ZOI_PLATFORM_SOURCE_DIR
+    )
+  # Trim a trailing whitespace.
+  string(STRIP ${ZOI_PLATFORM_SOURCE_DIR} ZOI_PLATFORM_SOURCE_DIR)
+
+  # Include ZBOSS build CMake file.
+  if(EXISTS "${ZOI_PLATFORM_SOURCE_DIR}/CMakeLists.txt")
+    include("${ZOI_PLATFORM_SOURCE_DIR}/CMakeLists.txt")
+  else()
+    message( FATAL_ERROR "Unable to find ZOI platform repository. Please checkout repositories from ZOI manifest file to recover. Exiting." )
+  endif()
```

Comment:
@tejlmand Only if they are members of ZBOSS Open Initative. What we'd like to achieve is to have the possibility to use the same NCS source for either library or source ZBOSS builds. This is simpler for users and simpler for us (maintenance, CI, etc.). 

I'll let @tomchy comment on the modules feature but as far as I remember we didn't see an elegant way to pull in the ZBOSS sources conditionally, i.e., give the user option to use ZBOSS source and its Kconfig when available. Do you think what you propose will help?

Reply:
@tejlmand as you may anticipated, we did not find the solution to make both - conditional `Kconfg` and `CMakeLists.txt` includes. The Zephyr modules feature allows us to do that (the ZOI main manifest repository may be a module), we have missed that.

This leaves us with two remaining issues:
- We still have to provide Kconfig options, that were available during stack compilation, so the common set of headers will select correct ZBOSS defines (please take a look at the `{nrfxlib_zboss}/include/osif/zb_vendor.h` file). This is the reason behind using `visible if ZBOSS_LIBRARY_SOURCES` instead of a simple `if` inside `Kconfig.sources` file. My proposition to tackle this issue is to export relevant Kconfig defines to a separate file as an artifact of ZBOSS build command and add the following line inside nrfxlib to pull it in:
```
toolchain_cc_imacros(${CMAKE_CURRENT_SOURCE_DIR}/include/osif/libzboss_config.h)
```
- We still have to conditionally remove CMakeLists commands under `if (CONFIG_ZBOSS_LIBRARY_PRECOMPILED)`. If we go by the path of removing all ZBOSS-source-related things from nrfxlib, the `Kconfig` file becomes empty and we need to add something to conditionally exclude lines from `CMakeLists.txt`, without Kconfig options. My proposition here is to use the path of the module as the switch:
```
if ((NOT DEFINED ${ZEPHYR_ZOI_MODULE_DIR}) OR
    (CONFIG_ZBOSS_LIBRARY_PRECOMPILED))
```

Additional note: using the Kconfig inside both nrfxlib and defining ZOI module will make the configuration counter-intuitive, with the following structure:
```
Zephyr Kernel Configuration
|
+ Modules. --->
|  |
|  + ZBOSS stack configuration.  --->
|  |
|  + Nordic nrfxlib -->
|  |   |
|  |   + ZBOSS stack configuration.  --->
|  |
|  + Nordic nRF Connect  --->
|     |
|     + Enable Zigbee stack  --->
```
This is the main reason, why I'm voting for making the nrfxlib's `Kconfig` file empty if the source-related options are removed from nrfxlib.

What do you think about my proposals?

Reply:
Regarding:
> We still have to provide Kconfig options, that were available during stack compilation, so the common set of headers will select correct ZBOSS defines (please take a look at the {nrfxlib_zboss}/include/osif/zb_vendor.h file). This is the reason behind using visible if ZBOSS_LIBRARY_SOURCES instead of a simple if inside Kconfig.sources file. My proposition to tackle this issue is to export relevant Kconfig defines to a separate file as an artifact of ZBOSS build command and add the following line inside nrfxlib to pull it in:

This appears very similar to a mbedTLS config file.
You could generated this file in same fashion as is done for mbedTLS, based on the Kconfig settings when you compile from source, and provide a static conf file that is part of the nrfxlib.
(i'm under the impression that this header file is static for users using the precompiled lib, correct ?)

Reply:
`Kconfig` pruned from source-code related options, `zb_vendor.h` left as-is, `libzboss_config.h` generated by the `config_file` command.